apps/src/blockly/addons/functionEditor.js

```diff
@@ -1,20 +1,108 @@
     opt_paramTypes
-  ) {}
+  ) {
+    // TODO: Are these options from the fork still relevant?
+    this.msgOverrides_ = opt_msgOverrides || {};
+    if (opt_definitionBlockType) {
+      this.definitionBlockType = opt_definitionBlockType;
+    }
+    this.parameterBlockTypes = opt_parameterBlockTypes || {};
+    this.disableParamEditing = opt_disableParamEditing || false;
+    this.paramTypes = opt_paramTypes || [];
+  }
 
-  // TODO
-  isOpen() {
-    return false;
+  setMainWorkspace = workspace => {
+    this.mainWorkspace = workspace;
+  };
+
+  setProcedureWorkspace = workspace => {
+    this.procedureWorkspace = workspace;
+  };
+
+  init(mainWorkspace, procedureWorkspace, options) {
+    this.setMainWorkspace(mainWorkspace);
+    this.setProcedureWorkspace(procedureWorkspace);
+
+    // The workspace we'll show to users for editing
+    const modalEditor = document.getElementById(MODAL_EDITOR_ID);
+    this.dom = modalEditor;
+
+    // Customize auto-populated Functions toolbox category.
+    this.editorWorkspace = Blockly.blockly_.inject(modalEditor, {
+      toolbox: options.toolbox,
+      theme: Blockly.cdoUtils.getUserTheme(options.theme),
+    });
+
+    // Close handler
+    document
+      .getElementById(MODAL_EDITOR_CLOSE_ID)
+      .addEventListener('click', () => this.hide());
+
+    // Rename handler
+    this.nameInput = document.getElementById(MODAL_EDITOR_NAME_INPUT_ID);
+    this.nameInput.addEventListener('input', e => {
+      this.block.getProcedureModel().setName(e.target.value);
+    });
+
+    // Description handler
+    this.functionDescriptionInput = document.getElementById(
+      'functionDescriptionText'
+    );
+    this.functionDescriptionInput.addEventListener('input', () => {
+      // TODO: Save the description to the procedure model
+    });
+
+    // Delete handler
+    document
+      .getElementById(MODAL_EDITOR_DELETE_ID)
+      .addEventListener('click', this.handleDelete.bind(this));
+
+    // Main workspace toolbox procedure category callback
+    this.mainWorkspace.registerToolboxCategoryCallback('PROCEDURE', () =>
+      this.flyoutCategory(this.mainWorkspace, true)
+    );
+
+    // Editor workspace toolbox procedure category callback
+    // we have to pass the main ws so that the correct procedures are populated
+    // false to not show the new function button inside the modal editor
+    this.editorWorkspace.registerToolboxCategoryCallback('PROCEDURE', () =>
+      this.flyoutCategory(this.mainWorkspace, false)
+    );
+
+    // Set up the "new procedure" button in the toolbox
+    this.mainWorkspace.registerButtonCallback('newProcedureCallback', () => {
+      this.newProcedureCallback(this.mainWorkspace);
+      // refresh the flyout after the new procedure is created
+      this.mainWorkspace.getToolbox().refreshSelection();
+    });
+
+    this.setUpEditorWorkspaceChangeListeners();
+  }
+
+  hide() {
+    this.dom.style.display = 'none';
+  }
+
+  hideIfOpen() {
```

Comment:
I don't think we need `hideIfOpen` and `hide` and I prefer the simpler name `hide`, but the reason I left this function in here is because of code in the following two places:
1. https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/StudioApp.js#L547 
2. https://github.com/code-dot-org/code-dot-org/blob/ab7bc270cbe272a430d86558af2c1deb56fd9615/apps/src/StudioApp.js#L839

This might be outside of the scope of this PR, but if we wanted to remove this TODO, we could either rename `hide` to `hideIfOpen` or rename the above usages to `hide`, right?

Reply:
yes I think the contract editor can go away once we've fully deprecated calc and eval (I think those projects will still show up?). We can rename the function to just hide, but in order to do that we would need to change the cdo version as well. So for now we might want to maintain both and just have `hideIfOpen` call `hide`.